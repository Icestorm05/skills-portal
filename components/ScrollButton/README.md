# Scroll Button

## Functionality

The scroll button component displays a fab, which scrolls to and from specified elements.

### Props

`<scroll-button>` supports the following custom component attributes:

| attribute | type | description
| --- | --- | ---
| `bottom` | String | The element to scroll to when at the top of the page.
| `top` | String | The element to scroll to when not at the top of the page.
| `offset` | Number *optional* | The scroll offset. This can be useful when there is a fixed element on the page. Defaults to `-150`.

### Example

```html
<div>
    <nav id="nav" height="100" style="position: fixed;">Navigation</nav>
    <div id="top" height="1000">Top Content</div>
    <div class="bottom" height="1000">Bottom Content</div>
    <scroll-button bottom=".bottom" top="#top" :offset="-100"/>
</div>
```

```ts
import Vue from 'vue';
import Component from 'nuxt-class-component';
import ScrollButton from '~/components/ScrollButton/ScrollButton.vue';
@Component({
    components: {
        ScrollButton,
    },
})
export default class GenericComponent extends Vue {}
```