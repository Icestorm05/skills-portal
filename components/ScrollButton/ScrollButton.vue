<template lang="pug">
    v-fab-transition
        v-btn(color="primary"
              fab
              fixed
              bottom
              left
              :aria-label="pageYOffset ? 'Scroll Up' : 'Scroll Down'"
              v-scroll-to="{ el: pageYOffset ? top : bottom, offset }").ScrollButton
            v-icon(v-if="pageYOffset").ScrollButtonIcon keyboard_arrow_up
            v-icon(v-else).ScrollButtonIcon keyboard_arrow_down
</template>

<script lang="ts">
import Component from "nuxt-class-component";
import Vue from "vue";
import { Prop } from "vue-property-decorator";

import { VFabTransition } from "vuetify/es5/components/transitions";
import VBtn from "vuetify/es5/components/VBtn";
import VIcon from "vuetify/es5/components/VIcon";

@Component({
  components: {
    VBtn,
    VFabTransition,
    VIcon
  }
})
export default class ScrollButton extends Vue {
  @Prop({ type: String, required: true })
  private bottom: string;
  @Prop({ type: String, required: true })
  private top: string;
  @Prop({ type: Number, default: -150 })
  private offset: number;

  private pageYOffset: number = window.pageYOffset;
  private mounted() {
    window.addEventListener("scroll", (e: Event) => {
      this.pageYOffset = window.pageYOffset;
    });
  }
}
</script>