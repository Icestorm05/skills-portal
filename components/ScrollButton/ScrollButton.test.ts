import ScrollButton from "./ScrollButton.vue";

import VueScrollTo from "vue-scrollto";
import { createLocalVue, mount } from "@vue/test-utils";

const localVue = createLocalVue();
localVue.use(VueScrollTo);

declare const expect: jest.Expect;

describe("Scroll Button", () => {
  test("is a Vue instance", () => {
    const wrapper = mount(ScrollButton, {
      localVue,
      propsData: { top: "body", bottom: "body" }
    });
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  test("snapshot has not changed", () => {
    const wrapper = mount(ScrollButton, {
      localVue,
      propsData: { top: "body", bottom: "body" }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("by default displays keyboard_arrow_down icon", () => {
    const wrapper = mount(ScrollButton, {
      localVue,
      propsData: { top: "body", bottom: "body" }
    });
    expect(wrapper.find(".ScrollButtonIcon").text()).toEqual(
      "keyboard_arrow_down"
    );
  });

  test("by default pageYOffset is 0", () => {
    const wrapper = mount(ScrollButton, {
      localVue,
      propsData: { top: "body", bottom: "body" }
    });
    expect(wrapper.vm.$data.pageYOffset).toEqual(0);
  });

  test("adds a scroll event listener on mount", () => {
    const spy = jest.spyOn(window, "addEventListener");
    const wrapper = mount(ScrollButton, {
      localVue,
      propsData: { top: "body", bottom: "body" }
    });
    expect(spy).toHaveBeenCalled();
  });

  test("pageYOffset updates with scroll event", () => {
    const wrapper = mount(ScrollButton, {
      localVue,
      propsData: { top: "body", bottom: "body" }
    });
    Object.defineProperty(window, "pageYOffset", { value: 10 });
    const event = new UIEvent("scroll", { detail: 1, view: window });
    window.dispatchEvent(event);
    expect(wrapper.vm.$data.pageYOffset).toEqual(10);
  });
});
