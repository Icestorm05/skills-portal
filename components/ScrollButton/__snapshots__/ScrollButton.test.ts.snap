// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scroll Button snapshot has not changed 1`] = `
<button type="button" class="ScrollButton btn btn--bottom btn--floating btn--fixed btn--left primary" aria-label="Scroll Down">
    <div class="btn__content">
        <i aria-hidden="true" class="icon ScrollButtonIcon material-icons">keyboard_arrow_down</i>
    </div>
</button>
`;
